File: events/year_2016/day_20/day_20.py
"""
Advent of Code - Year 2016 - Day 20
https://adventofcode.com/2016/day/20
"""


def preprocessing(puzzle_input: str) -> set[tuple[int, int]]:
    """
    Converts puzzle input into a set of blacklisted IP ranges.

    Args:
        puzzle_input (str): Raw puzzle input containing IP ranges

    Returns:
        set[tuple[int, int]]: Set of tuples containing lower and upper bounds of blacklisted ranges
    """
    blacklist = set()
    for element in puzzle_input.splitlines():
        inf, sup = (int(item) for item in element.split("-"))
        blacklist.add((inf, sup))
    return blacklist


def solver(intervales: set[tuple[int, int]], max_val: int = 4_294_967_295) -> tuple[int, int]:
    """
    Find the first allowed IP address and count total allowed IPs within blocked intervals.

    Args:
        intervales: Set of blocked IP ranges as (start, end) tuples
        max_val: Maximum IP address value (default: 4,294,967,295)

    Returns:
        Tuple of (first_allowed_ip, total_allowed_count)

    Examples:
        >>> solver({(5, 8), (0, 2), (4, 7)}, 9)
        (3, 2)
        >>> solver({(0, 0), (2, 2)}, 4)
        (1, 3)
    """
    counter_ip = 0
    allowed_ip = next_allowed_ip(0, intervales)
    first_allowed_ip = allowed_ip
    while allowed_ip <= max_val:
        intervales = set((inf, sup) for (inf, sup) in intervales if sup > allowed_ip)
        min_ip = min(intervales, key=lambda item: item[0], default=[max_val + 1])[0]
        counter_ip += min_ip - allowed_ip
        allowed_ip = next_allowed_ip(min_ip, intervales)
    return first_allowed_ip, counter_ip


def next_allowed_ip(start: int, intervales: set[tuple[int, int]]) -> int:
    """
    Find the next allowed IP address starting from a given position.

    Args:
        start: Starting IP address to check
        intervales: Set of blocked IP ranges as (min, max) tuples (inclusive)

    Returns:
        First IP address >= start that is not in any blocked range

    Examples:
        >>> next_allowed_ip(0, {(0, 2), (5, 7)})
        3
        >>> next_allowed_ip(10, {(0, 2), (5, 7)})
        10
        >>> next_allowed_ip(1, {(0, 5), (6, 10)})
        11
    """
    for inf, sup in intervales:
        if start in range(inf, sup + 1):
            intervales = set((inf, sup) for (inf, sup) in intervales if sup > start)
            return next_allowed_ip(sup + 1, intervales)
    return start

File: events/year_2016/day_20/day_20.rs
aoc::main!();

fn preprocessing(puzzle_input: &str) -> Vec<RangeInclusive<u32>> {
    let blacklist: Vec<RangeInclusive<u32>> = puzzle_input
        .split(['\n', '-'])
        .map(|n| n.parse().unwrap())
        .tuples()
        .map(|(start, end)| start..=end)
        .sorted_by(|a, b| Ord::cmp(&a.start(), &b.start()))
        .collect();

    let mut allowed_ip = next_allowed_ip(0, &blacklist);
    let mut intervales = Vec::new();
    while allowed_ip < u32::MAX {
        let next_ip = blacklist
            .iter()
            .find_or_first(|r| *r.start() > allowed_ip)
            .unwrap_or(&(u32::MAX - 1..=u32::MAX))
            .start();
        intervales.push(allowed_ip..=*next_ip);
        allowed_ip = next_allowed_ip(*next_ip, &blacklist);
    }
    intervales
}

fn part_1(whitelist: Vec<RangeInclusive<u32>>) -> u32 {
    *whitelist.first().unwrap().start()
}

fn part_2(blacklist: Vec<RangeInclusive<u32>>) -> u32 {
    blacklist.iter().map(|r| r.end() - r.start()).sum()
}

fn next_allowed_ip(ip: u32, blacklist: &Vec<RangeInclusive<u32>>) -> u32 {
    if ip == u32::MAX {
        return ip;
    }
    for r in blacklist {
        if r.contains(&ip) {
            return next_allowed_ip(r.end().saturating_add(1), blacklist);
        }
    }
    ip
}
